KAFKA-17492 skip features with minVersion of 0 instead of replacing 0 with 1 when BrokerRegistrationRequest < 4

m1a2st opened the pull request:
Jira: https://issues.apache.org/jira/browse/KAFKA-17492

Due to the BrokerRegistrationRequest minVersion is setting from 0 to 1, thus `ClusterControlManger#processRegistrationFeature` check the version will throw an error, fix it by below suggestion 
1. make 3.9+ broker be able to register to 3.8 controller
skip features with minVersion of 0 instead of replacing 0 with 1 when BrokerRegistrationRequest < 4

~2. make 3.9+ controller be able to accept unknown features from future broker
skip check of unknown features. see following sample code~

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

chia7712 commented:
> make 3.9+ broker be able to register to 3.8 controller
skip features with minVersion of 0 instead of replacing 0 with 1 when BrokerRegistrationRequest < 4

noticed that I had raised a question for it:  https://github.com/apache/kafka/pull/16421#discussion_r1748904248

junrao commented on clients/src/main/java/org/apache/kafka/common/requests/BrokerRegistrationRequest.java:
```diff
@@ -46,21 +46,7 @@ public short oldestAllowedVersion() {
-            } else {
-                return new BrokerRegistrationRequest(data, version);
-            }
+            return new BrokerRegistrationRequest(data, version);
```
We need to skip features with minSupportedVersion of 0 in v3 or less.

We need to do the same in ApiVersionResponse.

We need to change the comment like the following in ApiVersionResponse.json and BrokerRegistrationRequest.json.

`Note: in v0-v3, features with MinSupportedVersion = 0 show up with MinSupportedVersion = 1.`

chia7712 replied:
for another: please update the following comment:
```
                // Workaround for KAFKA-17011: for BrokerRegistrationRequest versions older than 4,
                // translate minSupportedVersion = 0 to minSupportedVersion = 1.
```

cmccabe replied:
Sorry, I don't understand the change here. The correct change (that matches the PR description) would be to remove features that had a minVersion of 0 when handling older RPCs, right? But instead the code just seems to pass everything through, ignoring RPC version. What am I missing?

junrao reviewed (commented):
@m1a2st : Thanks for the PR. Left one comment.

chia7712 commented on metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java:
```diff
@@ -501,14 +501,16 @@ BrokerFeature processRegistrationFeature(
-                    ". The broker wants a version between " + feature.minSupportedVersion() + " and " +
-                    feature.maxSupportedVersion() + ", inclusive.");
-        }
+        finalizedFeatures.get(feature.name()).ifPresent(
```
@m1a2st please take a look at https://github.com/apache/kafka/pull/16421#discussion_r1750717947.  We don't need to remove the check of `ClusterControlManager`

cmccabe replied:
This change is incorrect. If the minVersion is 0, the feature must be set.

chia7712 commented on clients/src/main/java/org/apache/kafka/common/requests/BrokerRegistrationRequest.java:
```diff
@@ -47,14 +47,14 @@ public short oldestAllowedVersion() {
                     BrokerRegistrationRequestData.Feature feature = iter.next();
                     if (feature.minSupportedVersion() == 0) {
-                        feature.setMinSupportedVersion((short) 1);
+                        iter.remove();
```
How about using `removeIf`? 
```java
newData.features().removeIf(feature -> feature.minSupportedVersion() == 0);
```

chia7712 reviewed (commented):
@cmccabe thanks for updating this PR. I will verify this patch on 3.9

chia7712 commented:
Run the `transactions_upgrade_test.py`, `transactions_mixed_versions_test.py`, and `kraft_upgrade_test.py` with this patch and #17084 on my local. all pass. 

I will merge it if no failed tests.

cmccabe committed:
KAFKA-17492: skip KIP-584 features with minVersion of 0 in older RPCs

Some older clients will have deserialization problems if a KIP-584 feature's minimum supported
level is 0.  Therefore, when preparing an ApiVersionResponse for a version less than 4, we must
omit features with a minimum supported level of 0. This is reasonable to do because older software
could not handle setting these features to a level other than 0 in any case.

A similar problem applies to BrokerRegistrationRequest. Here we must omit features with minVersion
= 0 when the RPC version is less than 4.

Note that this fix supersedes an earlier compatibility preservation effort in KAFKA-17011.

cmccabe commented:
I did some cleanup and added the `removeIf` simplification that was suggested

junrao commented on clients/src/main/java/org/apache/kafka/common/requests/ApiVersionsResponse.java:
```diff
@@ -289,19 +289,17 @@ private static SupportedFeatureKeyCollection maybeFilterSupportedFeatureKeys(
-                // at versions less than 4, we must set the minimum version for these features
-                // to 1 rather than 0. See KAFKA-17011 for details.
-                key.setMinVersion((short) 1);
+                // at versions less than 4, we must omit these features. See KAFKA-17492.
```
We need to change the comment like the following in ApiVersionResponse.json and BrokerRegistrationRequest.json.

`Note: in v0-v3, features with MinSupportedVersion = 0 show up with MinSupportedVersion = 1.
`

chia7712 replied:
@m1a2st Could you please address @junrao comment? That includes following files.

https://github.com/apache/kafka/blob/b22e848ddef975785846af13685ee50ce40d4df9/clients/src/main/resources/common/message/ApiVersionsResponse.json#L50
https://github.com/apache/kafka/blob/trunk/clients/src/main/resources/common/message/BrokerRegistrationRequest.json#L51

junrao commented on core/src/test/scala/unit/kafka/server/ApiVersionsResponseIntegrationTest.scala:
```diff
@@ -72,4 +72,12 @@ class ApiVersionsResponseIntegrationTest extends BaseRequestTest {
     assertEquals(expectedMinVersion, key.minVersion())
   }
+
+  def assertFeatureMissing(
```
Could this be private?

junrao reviewed (commented):
@cmccabe : Thanks for the updated PR. A couple of minor comments.

m1a2st committed:
addressed by comment

m1a2st commented:
Thanks for @cmccabe, @chia7712, @junrao helping for this issue, I will complete this PR

chia7712 commented on clients/src/main/java/org/apache/kafka/common/requests/ApiVersionsResponse.java:
```diff
@@ -289,19 +289,19 @@ private static SupportedFeatureKeyCollection maybeFilterSupportedFeatureKeys(
-                // at versions less than 4, we must set the minimum version for these features
-                // to 1 rather than 0. See KAFKA-17011 for details.
-                key.setMinVersion((short) 1);
+                // Note: For API versions v0-v3, features with MinSupportedVersion = 0 
```
This comment is not accurate. Could you consider using @cmccabe's comment?
```
                // Some older clients will have deserialization problems if a feature's
                // minimum supported level is 0. Therefore, when preparing ApiVersionResponse
                // at versions less than 4, we must omit these features. See KAFKA-17492.
```

chia7712 commented:
@m1a2st please add 3.8 version to following e2e:

https://github.com/apache/kafka/blob/trunk/tests/kafkatest/tests/core/zookeeper_migration_test.py#L89

chia7712 commented on clients/src/main/resources/common/message/ApiVersionsResponse.json:
```diff
@@ -47,7 +47,7 @@
     { "name":  "SupportedFeatures", "type": "[]SupportedFeatureKey", "ignorable": true,
       "versions":  "3+", "tag": 0, "taggedVersions": "3+",
-      "about": "Features supported by the broker. Note: in v0-v3, features with MinSupportedVersion = 0 show up with MinSupportedVersion = 1.",
+      "about": "Features supported by the broker. Note: in v0-v3, features blocked with MinSupportedVersion = 0.",
```
Pardon me, I don't catch "features blocked with MinSupportedVersion = 0."

Also,  please fix `BrokerRegistrationRequest.json` too

m1a2st replied:
> Pardon me, I don't catch "features blocked with MinSupportedVersion = 0."

I change it to omit, it is more accurancy in this case

chia7712 commented on clients/src/main/resources/common/message/BrokerRegistrationRequest.json:
```diff
@@ -48,7 +48,7 @@
     },
     { "name": "Features", "type": "[]Feature",
-      "about": "The features on this broker. Note: in v0-v3, features with MinSupportedVersion = 0 show up with MinSupportedVersion = 1.", "versions": "0+", "fields": [
+      "about": "The features on this broker. Note: in v0-v3, features omit with MinSupportedVersion = 0.", "versions": "0+", "fields": [
```
How about "features with MinSupportedVersion = 0 are omitted"?

chia7712 commented on clients/src/main/resources/common/message/ApiVersionsResponse.json:
```diff
@@ -47,7 +47,7 @@
     { "name":  "SupportedFeatures", "type": "[]SupportedFeatureKey", "ignorable": true,
       "versions":  "3+", "tag": 0, "taggedVersions": "3+",
-      "about": "Features supported by the broker. Note: in v0-v3, features with MinSupportedVersion = 0 show up with MinSupportedVersion = 1.",
+      "about": "Features supported by the broker. Note: in v0-v3, features omit with MinSupportedVersion = 0.",
```
How about "features with MinSupportedVersion = 0 are omitted"?

chia7712 reviewed (commented):
@m1a2st thanks for your patch. two minor comments are left. PTAL

chia7712 commented:
The failed tests pass on my local, and following links are the related jira. The new e2e works on my local.  I will merge it to trunk and 3.9 later if no objections.

https://issues.apache.org/jira/browse/KAFKA-17514
https://issues.apache.org/jira/browse/KAFKA-16174
https://issues.apache.org/jira/browse/KAFKA-15104
https://issues.apache.org/jira/browse/KAFKA-14533
https://issues.apache.org/jira/browse/KAFKA-17515
https://issues.apache.org/jira/browse/KAFKA-15772
https://issues.apache.org/jira/browse/KAFKA-15411
https://issues.apache.org/jira/browse/KAFKA-16501
https://issues.apache.org/jira/browse/KAFKA-15146
https://issues.apache.org/jira/browse/KAFKA-16377

cmccabe reviewed (approved):
LGTM. Thanks @m1a2st @junrao @chia7712

jolshan commented:
Thanks folks for fixing this 🙏

junrao commented on clients/src/main/resources/common/message/ApiVersionsResponse.json:
```diff
@@ -47,7 +47,7 @@
     { "name":  "SupportedFeatures", "type": "[]SupportedFeatureKey", "ignorable": true,
       "versions":  "3+", "tag": 0, "taggedVersions": "3+",
-      "about": "Features supported by the broker. Note: in v0-v3, features with MinSupportedVersion = 0 show up with MinSupportedVersion = 1.",
+      "about": "Features supported by the broker. Note: in v0-v3, features MinSupportedVersion = 0 are omitted.",
```
features MinSupportedVersion = 0 are omitted => features with MinSupportedVersion = 0 are omitted

junrao commented on clients/src/main/resources/common/message/BrokerRegistrationRequest.json:
```diff
@@ -48,7 +48,7 @@
     },
     { "name": "Features", "type": "[]Feature",
-      "about": "The features on this broker. Note: in v0-v3, features with MinSupportedVersion = 0 show up with MinSupportedVersion = 1.", "versions": "0+", "fields": [
+      "about": "The features on this broker. Note: in v0-v3, features MinSupportedVersion = 0 are omitted.", "versions": "0+", "fields": [
```
features MinSupportedVersion = 0 are omitted => features with MinSupportedVersion = 0 are omitted

junrao reviewed (commented):
@m1a2st : Thanks for the updated PR. Just a couple of minor comments.

m1a2st committed:
update by comments

junrao reviewed (approved):
@m1a2st : Thanks for the updated PR. LGTM

junrao commented:
@m1a2st : Could we update https://cwiki.apache.org/confluence/display/KAFKA/KIP-1022%3A+Formatting+and+Updating+Features and the mailing list with the changes in this PR?

```
In order to resolve this dilemma, we will create a new version 4 for ApiVersionsRequest. Clients which send v4 promise to be able to handle ranges including 0. Clients which send v3 will never see a range that starts with 0. Such ranges will be translated into ranges starting with 1 instead.

Similarly, for BrokerRegistrationRequest, we will create a new version 4. Servers which accept v4 promise to be able to handle ranges including 0. Servers which do not will get ranges that only start with 1, rather than 0.
```

m1a2st commented:
I will update this KIP https://cwiki.apache.org/confluence/display/KAFKA/KIP-1022%3A+Formatting+and+Updating+Features
